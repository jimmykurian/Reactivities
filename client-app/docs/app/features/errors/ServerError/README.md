[**Client App Documentation v0.0.0**](../../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../../README.md) / app/features/errors/ServerError

# app/features/errors/ServerError

## Index

### Functions

- [default](functions/default.md)
