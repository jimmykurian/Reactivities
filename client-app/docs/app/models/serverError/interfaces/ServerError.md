[**Client App Documentation v0.0.0**](../../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../../README.md) / [app/models/serverError](../README.md) / ServerError

# Interface: ServerError

The ServerError interface represents the structure of an error returned from the server.
It includes the HTTP status code, an error message, and optional detailed information about the error.

 ServerError

## Properties

### details

> **details**: `string`

Additional details or stack trace information about the error (optional).

#### Source

[src/app/models/serverError.ts:20](https://github.com/jimmykurian/Reactivities/blob/dbc3ed866e1f645e56a07b30e597ad4448fbda7a/client-app/src/app/models/serverError.ts#L20)

***

### message

> **message**: `string`

A brief message describing the error.

#### Source

[src/app/models/serverError.ts:19](https://github.com/jimmykurian/Reactivities/blob/dbc3ed866e1f645e56a07b30e597ad4448fbda7a/client-app/src/app/models/serverError.ts#L19)

***

### statusCode

> **statusCode**: `number`

The HTTP status code associated with the error.

#### Source

[src/app/models/serverError.ts:18](https://github.com/jimmykurian/Reactivities/blob/dbc3ed866e1f645e56a07b30e597ad4448fbda7a/client-app/src/app/models/serverError.ts#L18)
