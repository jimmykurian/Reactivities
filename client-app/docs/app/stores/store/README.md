[**Client App Documentation v0.0.0**](../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../README.md) / app/stores/store

# app/stores/store

## Index

### Interfaces

- [Store](interfaces/Store.md)

### Variables

- [StoreContext](variables/StoreContext.md)
- [store](variables/store.md)

### Functions

- [useStore](functions/useStore.md)
