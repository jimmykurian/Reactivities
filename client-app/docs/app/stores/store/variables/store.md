[**Client App Documentation v0.0.0**](../../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../../README.md) / [app/stores/store](../README.md) / store

# Variable: store

> `const` **store**: [`Store`](../interfaces/Store.md)

## Constant

store

## Description

The store instance containing all the application's stores.

## Source

[src/app/stores/store.ts:25](https://github.com/jimmykurian/Reactivities/blob/5670213c338d2af4595c1ce87506b162bb3099b4/client-app/src/app/stores/store.ts#L25)
