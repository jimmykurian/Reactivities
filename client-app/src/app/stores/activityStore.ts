/**
 * @file activityStore.tsx
 * @author Jimmy Kurian
 * @fileoverview This file contains the ActivityStore class which manages the state of activities using MobX.
 */

import { makeAutoObservable, runInAction } from 'mobx';
import { Activity } from '../models/activity';
import agent from '../api/agent';
import { v4 as uuid } from 'uuid';

/**
 * @class ActivityStore
 * @classdesc This class represents the MobX store for managing activity-related state.
 * It initializes the observables and configures MobX for state management.
 *
 * @example
 * ```tsx
 * import ActivityStore from './ActivityStore';
 *
 * const activityStore = new ActivityStore();
 * activityStore.loadActivities();
 * ```
 */
export default class ActivityStore {
  /** @type {Map<string, Activity>} */
  activityRegistry: Map<string, Activity> = new Map<string, Activity>();

  /** @type {Activity | undefined} */
  selectedActivity: Activity | undefined = undefined;

  /** @type {boolean} */
  editMode: boolean = false;

  /** @type {boolean} */
  loading: boolean = false;

  /** @type {boolean} */
  loadingInitial: boolean = true;

  /**
   * @constructor ActivityStore
   * @description Creates an instance of ActivityStore.
   * Initializes the observables using MobX's makeAutoObservable.
   */
  constructor() {
    makeAutoObservable(this);
  }

  /**
   * @getter activitiesByDate
   * @description Gets the activities sorted by date.
   * @returns {Activity[]}
   */
  get activitiesByDate(): Activity[] {
    return Array.from(this.activityRegistry.values()).sort(
      (a, b) => Date.parse(a.date) - Date.parse(b.date),
    );
  }

  /**
   * @async
   * @function loadActivities
   * @description Loads activities from the API and sets the activities state.
   * @returns {Promise<void>}
   */
  loadActivities = async (): Promise<void> => {
    try {
      const activities = await agent.Activities.list();
      runInAction(() => {
        activities.forEach((activity: Activity) => {
          activity.date = activity.date.split('T')[0];
          this.activityRegistry.set(activity.id, activity);
        });
        this.setLoadingInitial(false);
      });
    } catch (error) {
      console.log(error);
      runInAction(() => {
        this.setLoadingInitial(false);
      });
    }
  };

  /**.
   * @function setLoadingInitial
   * @description Sets the loading initial state
   * @param {boolean} state - The loading state to set.
   */
  setLoadingInitial = (state: boolean): void => {
    this.loadingInitial = state;
  };

  /**
   * @function selectActivity
   * @description Selects an activity by its ID.
   * @param {string} id - The ID of the activity to select.
   */
  selectActivity = (id: string): void => {
    this.selectedActivity = this.activityRegistry.get(id);
  };

  /**
   * @function cancelSelectedActivity
   * @description Cancels the selection of the currently selected activity.
   */
  cancelSelectedActivity = (): void => {
    this.selectedActivity = undefined;
  };

  /**
   * @function openForm
   * @description Opens the form for creating or editing an activity.
   * @param {string} [id] - The ID of the activity to edit, if any.
   */
  openForm = (id?: string): void => {
    id ? this.selectActivity(id) : this.cancelSelectedActivity();
    this.editMode = true;
  };

  /**
   * @function closeForm
   * @description Closes the form for creating or editing an activity.
   */
  closeForm = (): void => {
    this.editMode = false;
  };

  /**
   * @async
   * @function createActivity
   * @description Creates a new activity.
   * @param {Activity} activity - The activity to create.
   * @returns {Promise<void>}
   */
  createActivity = async (activity: Activity): Promise<void> => {
    this.loading = true;
    activity.id = uuid();
    try {
      await agent.Activities.create(activity);
      runInAction(() => {
        this.activityRegistry.set(activity.id, activity);
        this.selectedActivity = activity;
        this.editMode = false;
        this.loading = false;
      });
    } catch (error) {
      console.log(error);
      runInAction(() => {
        this.loading = false;
      });
    }
  };

  /**
   * @async
   * @function updateActivity
   * @description Updates an existing activity.
   * @param {Activity} activity - The activity to update.
   * @returns {Promise<void>}
   */
  updateActivity = async (activity: Activity): Promise<void> => {
    this.loading = true;
    try {
      await agent.Activities.update(activity);
      runInAction(() => {
        this.activityRegistry.set(activity.id, activity);
        this.selectedActivity = activity;
        this.editMode = false;
        this.loading = false;
      });
    } catch (error) {
      console.log(error);
      runInAction(() => {
        this.loading = false;
      });
    }
  };

  /**
   * @async
   * @function deleteActivity
   * @description Deletes an activity by its ID.
   * @param {string} id - The ID of the activity to delete.
   * @returns {Promise<void>}
   */
  deleteActivity = async (id: string): Promise<void> => {
    this.loading = true;
    try {
      await agent.Activities.delete(id);
      runInAction(() => {
        this.activityRegistry.delete(id);
        if (this.selectedActivity?.id === id) this.cancelSelectedActivity();
        this.loading = false;
      });
    } catch (error) {
      console.log(error);
      runInAction(() => {
        this.loading = false;
      });
    }
  };
}
