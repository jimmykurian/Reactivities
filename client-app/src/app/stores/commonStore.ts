/**
 * @file CommonStore.ts
 * @author Jimmy Kurian
 * @fileoverview This file contains the MobX store class `CommonStore`, which is responsible for managing common application state,
 * specifically handling server errors throughout the Reactivities application.
 */

import { makeAutoObservable } from 'mobx';
import { ServerError } from '../models/serverError';

/**
 * CommonStore class manages common application-wide state, including server error handling.
 * This store is made observable using MobX, allowing components to reactively update based on changes to the store's state.
 *
 * @class
 */
export default class CommonStore {
  /**
   * Represents the server error state, initially set to null.
   * The `ServerError` model contains the error message and optional stack trace details.
   *
   * @type {ServerError | null}
   */
  error: ServerError | null = null;

  /**
   * Initializes a new instance of the `CommonStore` class and makes it observable using MobX.
   */
  constructor() {
    makeAutoObservable(this);
  }

  /**
   * Sets the server error state in the store.
   *
   * @param {ServerError} error - The server error object containing details of the error.
   */
  setServerError = (error: ServerError) => {
    this.error = error;
  };
}
