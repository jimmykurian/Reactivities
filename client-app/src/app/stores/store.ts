/**
 * @file store.ts
 * @author Jimmy Kurian
 * @fileoverview This file contains the store for the application.
 */

import { createContext, useContext } from 'react';
import ActivityStore from './activityStore';
import CommonStore from './commonStore';

/**
 * @interface Store
 * @description interface representing the structure of the store.
 *
 * @property {ActivityStore} activityStore - Instance of ActivityStore for managing activity-related state.
 */
export interface Store {
  activityStore: ActivityStore;
  commonStore: CommonStore;
}

/**
 * @constant store
 * @description The store instance containing all the application's stores.
 * @type {Store}
 */
export const store: Store = {
  activityStore: new ActivityStore(),
  commonStore: new CommonStore(),
};

/**
 * @constant StoreContext
 * @description The React context for the store.
 * @type {React.Context<Store>}
 */
export const StoreContext: React.Context<Store> = createContext(store);

/**
 * @function Store
 * @description Custom hook to use the store in React components.
 * @returns {Store} The store instance.
 * @example
 * ```tsx
 * import { useStore } from './store';
 *
 * const MyComponent = () => {
 *   const { activityStore } = useStore();
 *   return <div>{activityStore.title}</div>;
 * };
 * ```
 */
export function useStore(): Store {
  return useContext(StoreContext);
}
