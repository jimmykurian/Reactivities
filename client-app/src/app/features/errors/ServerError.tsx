/**
 * @file ServerError.tsx
 * @author Jimmy Kurian
 * @fileoverview This file contains the ServerError component, which is responsible for displaying server-related errors to the user.
 * The component fetches the error details from the MobX store and displays them with relevant headers and formatting.
 */

import { observer } from 'mobx-react-lite';
import { useStore } from '../../stores/store';
import { Container, Header, Segment } from 'semantic-ui-react';

/**
 * ServerError component displays a server error message and optionally shows the error's stack trace details.
 * It retrieves the error information from the MobX `commonStore` and renders it in a user-friendly format.
 *
 * @component
 * @returns {JSX.Element} The rendered ServerError component displaying error details and stack trace if available.
 *
 * @example
 * ```tsx
 * import ServerError from './ServerError';
 *
 * function App() {
 *   return <ServerError />;
 * }
 * ```
 */
export default observer(function ServerError(): JSX.Element {
  const { commonStore } = useStore();

  return (
    <Container>
      <Header as="h1" content="Server Error" />
      <Header sub as="h5" color="red" content={commonStore.error?.message} />
      {commonStore.error?.details && (
        <Segment>
          <Header as="h4" content="Stack trace" color="teal" />
          <code style={{ marginTop: '10px' }}>{commonStore.error.details}</code>
        </Segment>
      )}
    </Container>
  );
});
