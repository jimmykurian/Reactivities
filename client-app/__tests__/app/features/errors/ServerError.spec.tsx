import { render, screen } from '@testing-library/react';
import { faker } from '@faker-js/faker';
import { useStore } from '../../../../src/app/stores/store';
import ServerError from '../../../../src/app/features/errors/ServerError';

jest.mock('../../../../src/app/stores/store', () => ({
  useStore: jest.fn(),
}));

describe('ServerError Component', () => {
  const staticErrorData = {
    error: {
      message: 'Static server error occurred',
      details: 'Static stack trace details for the error',
    },
  };

  const mockStore = (error: unknown) => {
    (useStore as jest.Mock).mockReturnValue({
      commonStore: {
        error,
      },
    });
  };

  it('should display error message when error exists', () => {
    // Arrange
    const errorData = {
      message: faker.lorem.sentence(),
    };
    mockStore(errorData);

    // Act
    render(<ServerError />);

    // Assert
    expect(screen.getByText('Server Error')).toBeInTheDocument();
    expect(screen.getByText(errorData.message)).toBeInTheDocument();
  });

  it('should display stack trace details if they exist', () => {
    // Arrange
    const errorData = {
      message: faker.lorem.sentence(),
      details: faker.lorem.paragraph(),
    };
    mockStore(errorData);

    // Act
    render(<ServerError />);

    // Assert
    expect(screen.getByText('Stack trace')).toBeInTheDocument();
    expect(screen.getByText(errorData.details)).toBeInTheDocument();
  });

  it('should not display stack trace section if there are no details', () => {
    // Arrange
    const errorData = {
      message: faker.lorem.sentence(),
      details: null,
    };
    mockStore(errorData);

    // Act
    render(<ServerError />);

    // Assert
    expect(screen.queryByText('Stack trace')).not.toBeInTheDocument();
  });

  it('should render correctly with static data (snapshot)', () => {
    // Arrange
    mockStore(staticErrorData);

    // Act
    const { container } = render(<ServerError />);

    // Assert
    expect(container).toMatchSnapshot();
  });

  it('should render "Server Error" message even when there is no error', () => {
    // Arrange
    mockStore(null);

    // Act
    render(<ServerError />);

    // Assert
    expect(screen.getByText('Server Error')).toBeInTheDocument();
    expect(screen.queryByText('Stack trace')).not.toBeInTheDocument();
  });
});
