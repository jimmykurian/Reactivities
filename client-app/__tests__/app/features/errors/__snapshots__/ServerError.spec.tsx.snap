// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServerError Component should render correctly with static data (snapshot) 1`] = `
<div>
  <div
    class="ui container"
  >
    <h1
      class="ui header"
    >
      Server Error
    </h1>
    <h5
      class="ui red sub header"
    />
  </div>
</div>
`;
